python/mead/downloader.py

```diff
@@ -0,0 +1,300 @@
+    return path_to_save
+
+
+def validate_url(url):
+    import re
+    regex = re.compile(
+        r'^(?:http|ftp)s?://' # http:// or https://
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|' #domain...
+        r'localhost|' #localhost...
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})' # ...or ip
+        r'(?::\d+)?' # optional port
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+    return re.match(regex, url) is not None
+
+
+def update_cache(key):
+    import json
+    try:
+        dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))[key]
+    except (IOError, KeyError):
+        return
+    del dcaches[key]
+    json.dump(dcaches, open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json"), "w"), indent=True)
+
+
+def delete_file(file_loc):
+    import os
+    import shutil
+    if os.path.exists(file_loc):
+        shutil.rmtree(file_loc)
+
+
+def check_sanity_file(file_loc, key=None):
+    import os
+    if os.path.exists(file_loc) and os.path.isfile(file_loc) and not mime_type(file_loc) == "text/html":  # dropbox doesn't give 404 in case the file does not exist, produces an HTML. Almost never the actual train/dev/test file is an HTML.
+        return True
+    else:
+        delete_file(file_loc)
+        if key is not None: # cache file validation
+            update_cache(key)
+        return False
+
+
+def check_sanity_dir(dir_loc, dataset_desc, ignore_file_check=False, key=None):
+    import os
+    if not os.path.exists(dir_loc) or not os.path.isdir(dir_loc):
+        update_cache(dir_loc, key)
+        return False
+    if ignore_file_check:  # for enc_dec tasks there's no direct downloads
+        return True
+    files = [os.path.join(dir_loc, dataset_desc[k]) for k in dataset_desc if k.endswith("_file")]
+    for file in files:
+        if not check_sanity_file(file, key):
+            return False
+    return True
+
+
+class Downloader(object):
+    def __init__(self, data_download_cache, cache_ignore):
+        super(Downloader, self).__init__()
+        self.cache_ignore = cache_ignore
+        self.data_download_cache = data_download_cache
+        pass
+
+    def download(self):
+        pass
+
+
+class SingleFileDownloader(Downloader):
+    def __init__(self, dataset_file, data_download_cache, cache_ignore=False):
+        super(SingleFileDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.dataset_file = dataset_file
+        self.data_download_cache = data_download_cache
+
+    def download(self):
+        import json
+        file_loc = self.dataset_file
+        if check_sanity_file(file_loc):
+            return file_loc
+        elif validate_url(file_loc):  # is it a web URL? check if exists in cache
+            url = file_loc
+            try:
+                dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))
```

Comment:
fairly certain opening files requires a call to close() to free up system resources.